lib/src/main/java/org/esbtools/lightbluenotificationhook/NotificationHook.java

```diff
@@ -120,69 +120,61 @@ private HookResult processSingleHookDoc(EntityMetadata metadata,
             return HookResult.aborted();
         }
 
-        if (watchedFieldsHaveChanged(metadata,preDoc, postDoc, watchProjector, arrayOrderingSignificant)) {
-            LOGGER.debug("Watched fields changed, creating notification");
-            NotificationEntity notification =
-                makeNotificationEntityWithIncludedFields(hookDoc, includeProjector);
-
-            EntityVersion notificationVersion = new EntityVersion(
-                    NotificationEntity.ENTITY_NAME,
-                    NotificationEntity.ENTITY_VERSION);
-
-            InsertionRequest newNotification = new InsertionRequest();
-
-            newNotification.setEntityVersion(notificationVersion);
-            newNotification.setEntityData(objectMapper.valueToTree(notification));
-
-            LOGGER.debug("Inserting notification");
-            Response response = mediator.insert(newNotification);
-
-            return new HookResult(notification, response.getErrors(), response.getDataErrors());
+        try {
+            DocComparator.Difference<JsonNode> diff=compareDocs(metadata,preDoc,postDoc,watchProjector);
+            if(!diff.same()) {
+                if(diff.getNumChangedFields()>0 || arrayOrderingSignificant) {                
+                    LOGGER.debug("Watched fields changed, creating notification");
+                    NotificationEntity notification =
+                        makeNotificationEntityWithIncludedFields(hookDoc, includeProjector, diff, arrayOrderingSignificant);
+                    
+                    EntityVersion notificationVersion = new EntityVersion(NotificationEntity.ENTITY_NAME,
+                                                                          NotificationEntity.ENTITY_VERSION);
+                    
+                    InsertionRequest newNotification = new InsertionRequest();
+                    
+                    newNotification.setEntityVersion(notificationVersion);
+                    newNotification.setEntityData(objectMapper.valueToTree(notification));
+                    
+                    LOGGER.debug("Inserting notification");
+                    Response response = mediator.insert(newNotification);
+                    
+                    return new HookResult(notification, response.getErrors(), response.getDataErrors());
+                }
+            }
+        } catch (Exception e) {
+            LOGGER.error("Error processing hook:"+e);
```

Comment:
Changed exception processing, let me know if this looks good.

On Thu, Mar 3, 2016 at 12:08 PM, Alec Henninger <notifications@github.com>
wrote:

> In
> lib/src/main/java/org/esbtools/lightbluenotificationhook/NotificationHook.java
> <https://github.com/esbtools/lightblue-notification-hook/pull/7#discussion_r54930089>
> :
>
> > +                    EntityVersion notificationVersion = new EntityVersion(NotificationEntity.ENTITY_NAME,
> > +                                                                          NotificationEntity.ENTITY_VERSION);
> > +
> > +                    InsertionRequest newNotification = new InsertionRequest();
> > +
> > +                    newNotification.setEntityVersion(notificationVersion);
> > +                    newNotification.setEntityData(objectMapper.valueToTree(notification));
> > +
> > +                    LOGGER.debug("Inserting notification");
> > +                    Response response = mediator.insert(newNotification);
> > +
> > +                    return new HookResult(notification, response.getErrors(), response.getDataErrors());
> > +                }
> > +            }
> > +        } catch (Exception e) {
> > +            LOGGER.error("Error processing hook:"+e);
>
> I replied to an old thread about this, but we should capture this
> Exception with a HookResult IMO. We can do that by changing HookResult to
> work with Exception instead of the lightblue Error type and that should
> accommodate both cases.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/esbtools/lightblue-notification-hook/pull/7/files#r54930089>
> .
>